### PR TITLE
Huge update to new design

### PR DESCRIPTION
**Pull request by coszio**
Coauthored by @agourlay. Most of the work was done on pair programming sessions.

Updates everything to match the new bitmask design with headerless pages.

## TODO
- [x] introduce gaps structure to find empty spots quicker.

### CODE REVIEW

**Comment by coszio** [comment]
Just pushed the changes to add `Gaps` structure to each region. (Coauthored by @agourlay again).

However, there is a test that is now failing because we changed the logic to creating new pages, can add an extra page. 
We need to optimize this part to fix the test.

Additionally, although the `Gaps` structures improve performance, it is still quite a lot slower than previous implementation. Let's discuss ways to improve 2 things:
- replace `BitSlice::last_one` with something faster.
- put the `Gaps` in a different structure so we don't have to do a linear scan to find a fitting gap.

**Comment by coszio** [comment]
Optimization update for:
- `find_available_blocks`
- `calculate_gaps`

These two were slowest functions when profiling, and have been optimized to run about 10x faster.

They now process the bitslices in chunks of 64 bits to take advantage of bitwise operators over `usize`.

## Flamegraphs

Before optimizations
imagen


After optimizing `calculate_gaps`
imagen


After optimizing `find_available_blocks`
imagen

## Benchmarks
Before opts:

```**read_heavy** with prefill_fraction 0.95
PayloadStorage:
1572864 operations across 1 thread(s) in 4.020268542s; time/op = 2.556µs
RocksDB:
1572864 operations across 1 thread(s) in 937.348584ms; time/op = 595ns

**insert_heavy** with prefill_fraction 0.0
PayloadStorage:
1572864 operations across 1 thread(s) in 82.551443375s; time/op = 52.484µs
RocksDB:
1572864 operations across 1 thread(s) in 3.69300475s; time/op = 2.347µs

**update_heavy** with prefill_fraction 0.5
PayloadStorage:
1572864 operations across 1 thread(s) in 30.615338042s; time/op = 19.464µs
RocksDB:
1572864 operations across 1 thread(s) in 2.141997875s; time/op = 1.361µs
```


After opts:
```
**read_heavy** with prefill_fraction 0.95
PayloadStorage:
1572864 operations across 1 thread(s) in 1.23288875s; time/op = 783ns
RocksDB:
1572864 operations across 1 thread(s) in 847.860959ms; time/op = 539ns

**insert_heavy** with prefill_fraction 0.0
PayloadStorage:
1572864 operations across 1 thread(s) in 6.560875792s; time/op = 4.171µs
RocksDB:
1572864 operations across 1 thread(s) in 3.393870416s; time/op = 2.157µs

**update_heavy** with prefill_fraction 0.5
PayloadStorage:
1572864 operations across 1 thread(s) in 4.655892083s; time/op = 2.96µs
RocksDB:
1572864 operations across 1 thread(s) in 2.005825375s; time/op = 1.275µs
```

**TLDR;** It is still slower than RocksDB, but not on a different league.

I still believe we can make this faster by putting some more brainpower into this.

**Comment by timvisee** [comment]
> They now process the bitslices in chunks of 64 bits to take advantage of bitwise operators over usize.

I assume you're talking about using the `usize` store type in bit slices. Could you check what performance is like with these two configurations as well:
- [`Msb0`](https://docs.rs/bitvec/latest/bitvec/order/struct.Msb0.html) bit order and `u8` bit store
- [`Lsb0`](https://docs.rs/bitvec/latest/bitvec/order/struct.Lsb0.html) bit order and `usize` bit store

The above two allow portability across different architectures, but may have a slight performance impact.

Relevant docs: https://ferrilab.github.io/bitvec/memory-representation.html#little-endian-byte-ordered-machines

**Review by agourlay — APPROVED** [review]
LGTM
